### PR TITLE
Breadth First Search (BFS)

### PR DESCRIPTION
**Pull request by Rafeh01**
<h1>Breadth First Search (BFS)</h1>
<h3>A Searching Algorithm Based on Graph Traversal</h3>
<h6>Breadth First Search Vs. Depth First Search</h6>
<li> `BFS` uses a **queue** (FIFO) data structure in its implementation.
<li> `DFS` uses a **stack** (FILO) data structure in its implementation. </li>
1. I added the **implementation** of the breadth first search algorithm.
2. I added the **proper tests** for the breadth first search algorithm.

Please let me know if there is anything else that I need to add :smile:.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -7,3 +7,4 @@ build/
 *~
 /dist/
 /*.egg-info
+*config.py
```
Comment (nryoung) [inline .gitignore]: I am not sure why this is necessary?

File: .travis.yml [hunk 2]
```diff
@@ -7,7 +7,6 @@ env:
   - TOX_ENV=py32
   - TOX_ENV=py33
   - TOX_ENV=py34
-  - TOX_ENV=py35
```
Comment (nryoung) [inline .travis.yml]: I am not sure why this change was made?

**Comment by nryoung** [comment]
Hey, thanks for your contribution. A few comments:

* You have added a lot of misc files that look like are specific to your IDE. You will need to remove those from being tracked by git.
* In order to keep a clean commit history on master could you squash and rebase your commits? You can follow this guide: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

**Comment by Rafeh01** [comment]
**Disclaimer:** I have never done rebasing before. 
However, I am very excited to announce that after about 2 hours of solid work. Not only do I understanding rebasing much better now, I believe I also fixed the intial problem of having too many rebases. The best I could cut it down to was 3 commits. Let me know if there is something left for me to do. Thanks for your comments and by you enforcing a restriction, I learned a new amazing concept of rebasing in git!

**Comment by nryoung** [comment]
Glad you learned a new concept!

Can you please squash and rebase so that there is only one commit? Can you also simplify your commit message to only explain what you are adding/modifying? For example:

```
Add iterative breadth first search
```

**Comment by Rafeh01** [comment]
Done! I squashed and rebased it again. My branch is only 1 commit ahead and the commit is now named "add iterative breadth first search". Hopefully, that fixes it!

File: algorithms/searching/breadth_first_search.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+"""
+Author: Rafeh Qazi (Rafeh01 on github)
+Iterative queue based implementation of a BFS algorithm on a graph.
+Breadth First Search Overview:
+    Used to traverse graphs or trees.
+    Starts at a selected node and searches all the closest connected
+    nodes first before moving on to the next nodes.
+"""
```
Comment (nryoung) [inline algorithms/searching/breadth_first_search.py]: Can you match the doc strings found in other modules? For an example take a look at: https://github.com/nryoung/algorithms/blob/master/algorithms/sorting/strand_sort.py#L1-L18

**Comment by nryoung** [comment]
Just one further nitpick. I have made an inline comment.

**Comment by Rafeh01** [comment]
Done! Improved the documentation as you requested and also removed the docs from my function. It is now more to the standards of the other commits that were made and accepted previously.

**Comment by nryoung** [comment]
Looks good! Thanks for the hard work on this.
